### DISCUSSION TITLE
I'm getting an HTTP 403 error when trying to publish a package to our organization, even though I have write permissions

### DISCUSSION THREAD

**Discussion by Rod-at-DOH** [community/community#54975] [Enterprise] [answered]
### Select Topic Area

Question

### Body

I'm trying to publish for the first time a NuGet package to our organization. I've never done this before, so I've used several websites to get help, such as, "Quickstart: Create and publish a NuGet package using Visual Studio (Windows only)". In this case we're not using an SDK style project, as the project was written using .NET Framework 4.5.2. Therefore, on the referenced page I used the [NuGet CLI](https://learn.microsoft.com/en-us/nuget/quickstart/create-and-publish-a-package-using-visual-studio?tabs=nuget#publish-with-the-net-cli-or-nuget-cli). Here is the error message I got:

> Pushing SendMail.1.0.0.nupkg to 'https://nuget.pkg.github.com/OUR-Sandbox'...
  PUT https://nuget.pkg.github.com/OUR-Sandbox/
WARNING: Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured.
  Forbidden https://nuget.pkg.github.com/OUR-Sandbox/ 539ms
Response status code does not indicate success: 403 (Forbidden).

And here's the command I used in PowerShell:

> nuget.exe push .\SendMail.1.0.0.nupkg ghp_MyClassicToken -Source https://nuget.pkg.github.com/OUR-Sandbox

When I created my classic token, I created it so that it has read, write, update, delete privileges. So, why am I getting an HTTP 403 error?

**Comment by jge162** [comment]
Hello @Rod-at-DOH maybe you have not added to the NuGet package source as a valid authentication token. Below -> Replace `OWNER` with the name of the GitHub organization or `user account` that owns the repository, replace `USERNAME` with the username associated with the classic token, and replace `TOKEN` with the classic token itself. So after you run this in power shell try again and please report back after.

`nuget.exe sources add -name "GitHub Packages" -source "https://nuget.pkg.github.com/OWNER/index.json" -username USERNAME -password TOKEN`

**Reply by Rod-at-DOH** [reply]
Hello @jge162, thank you for your response. Just to clarify some things from your answer, I've got a few questions.

The `nuget.exe sources add` command looks like it's adding that source to something. What and where?

The username associated with the classic token will be mine, I presume I'd use Rod-at-DOH, correct? (Not my email address.)

**Reply by jge162** [reply]
Hello again @Rod-at-DOH it adds the GitHub Packages `source` to your list of available package `sources`. The `-source` parameter specifies the `URL` of the package `source`, and the `-name` parameter specifies a friendly name for the package source that will be displayed in the NuGet package manager. Additionally, you are correct about your presumption. 
> I presume I'd use Rod-at-DOH, correct? (Not my email address.)

**Reply by Rod-at-DOH** [reply]
Hello @jge162, that got me closer, but I'm still getting a HTTP 403 error. First, it timed out, so I did as the message suggested, by adding `-Timeout 600` to the `nuget.exe push` command. However, after increasing the timeout and re-running the command I got this error:

> Pushing SendMail.1.0.0.nupkg to 'https://nuget.pkg.github.com/OUR-Sandbox'...
  PUT https://nuget.pkg.github.com/OUR-Sandbox/
An error was encountered when fetching 'PUT https://nuget.pkg.github.com/OUR-Sandbox/'. The request will now be retried.
An error occurred while sending the request.
  The remote name could not be resolved: 'nuget.pkg.github.com'
  PUT https://nuget.pkg.github.com/OUR-Sandbox/
WARNING: Your request could not be authenticated by the GitHub Packages service. Please ensure your access token is valid and has the appropriate scopes configured.
  Forbidden https://nuget.pkg.github.com/OUR-Sandbox/ 17635ms
Response status code does not indicate success: 403 (Forbidden).

I'm certain my access token is valid. At least as certain as I can be. I created it yesterday. Is there a way I can view the token to see what permissions is has?

**Reply by jge162** [reply]
You can view token in settings under developer section. Seems like a permission issue.

**Reply by Rod-at-DOH** [reply]
Thank you, I've found the token in settings. It has write permissions to GH Packages.

So, it is likely a permissions issue? Where? At GH's level? Or our Enterprise Organization level? Or at our end? How do I determine where the problem is?

**Reply by BrycenDavisSH** [reply]
Did you ever figure out how to resolve this? We have a few developers that are having this same problem and everything on the tokens and the nuget sources all look correct.

**Reply by Rod-at-DOH** [reply]
No, I have not resolved this issue

**Reply by marlonbraga** [reply]
I had this problem. Check the name of your NUGET SERVER
It must contain the github account name and not the project name

--source "https://nuget.pkg.github.com/Rod-at-DOH/index.json"
`                                      ^^^^^^^^`

`dotnet nuget push bin/Debug/name-package.3.0.0.nupkg --api-key ************************************ --source "https://nuget.pkg.github.com/Rod-at-DOH/index.json" --skip-duplicate`

**Reply by Rod-at-DOH** [reply]
The package I'm trying to publish will go into an organization. Do I need to include the organization name as well as my GitHub account name in the source URL?

**Reply by BrycenDavisSH** [reply]
We were able to get past this. Check that the package is attached to a repo in your organization, that you have the Inherit Permissions set for the package and your user/group at least has 'read' access to the repo. Not sure why it needs access to both, but that at least seemed to solve our issues.

**Reply by aknosis** [reply]
@BrycenDavisSH How were you able to attach to a repo if you can't push with your token? There is no UI present for a package prior to being uploaded.

**Reply by BrycenDavisSH** [reply]
@aknosis For us, it was only affecting some of our users in the organization and not everyone, for some unknown reason. So we just had a user that it wasn't affecting push the packages and fix the permissions until all users could pull/push packages.

**Reply by aknosis** [reply]
@BrycenDavisSH Thanks for the response, that is what I was afraid of. Definitely a chicken and egg situation here.

**Reply by Rod-at-DOH** [reply]
Hi @jge162, @BrycenDavisSH, _et al_, I'm sorry it's taken me so long to get back to this thread, but other urgent projects have demanded my attention.

I'm trying to pick up this thread again. I've searched for additional help for getting a token in place, so I can publish one of the packages I mentioned back on May 9th, to the repo in our GitHub organization. I came across this GH help document [Managing your personal access tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens), which confuses me. I'm not trying to create a publishing token that will only be used by me, but by everyone who uses the repo in our organization, in the GH Action(s) they will write. I am misreading that GH document, or does it not even pertain to what I'm trying to do?

**Reply by aknosis** [reply]
In my case I am trying to use a [token generated by a GitHub App](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-an-installation-access-token-for-a-github-app) to upload packages. Apparently this does not work.

I was able to find a solution that fit my needs since my use case is inside of a GitHub Actions workflow run. Apparently the token generated specifically for that workflow run does have access to upload packages. So I set that token as an env variable and use it to upload the nuget package.

```yaml
jobs:
  job:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write
      packages: write
    steps:
      - name: Example
        env:
          REPO_SCOPED_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: python script.py
```

Inside the script:

```python
process = subprocess.Popen(
  ['dotnet', 'nuget', 'push', 'artifact', '--source', 'https://nuget.pkg.github.com/{org}/index.json', '--api-key', os.environ['REPO_SCOPED_GITHUB_TOKEN'], '--skip-duplicate']
)
```

The nice thing about this, is that the upload is already attributed to the repo where the workflow is running so I do not have to do any extra steps with permissions post upload, it already inherits from the repo.

**Reply by Rod-at-DOH** [reply]
Hi @jge162, I want to go back to the topic of _sources_ that you brought up back on May 9. No where in the tutorials I've found and followed, is _sources_ brought up. Why do I need to include a source? What **are** _sources_?

And since the topic of _sources_ wasn't brought up in the GitHub Packages tutorials I've found and followed, is there anything else I need to include, to publish a non-SDK DLL to a GitHub Package that's a part of a repository in our organization?

**Reply by moonwitch** [reply]
Hi @Rod-at-DOH 
I was wondering if you have found a solution, since I have a similar issue - or rather the same, but then with Powershell Publishing to GH Packages.

**Reply by Rod-at-DOH** [reply, marked as answer]
Yes @moonwitch, I have found a solution. It's somewhat complicated and it requires using an Enterprise account with an organization set up where I publish the package to. So, at this point I haven't published to a personal GH Repo Package; on the other hand, I wasn't trying to. As they say in Math class, "That exercise is left for the student." Here's the steps I took:

### Install NuGet
Make sure you're [installed NuGet](https://learn.microsoft.com/en-us/nuget/install-nuget-client-tools) (I had to, because I was working with .NET Framework DLLs, not the newer SDK style DLLs.)

### Set up the environment on your dev box
For reference, I used this Microsoft Learn course on [creating a package](https://learn.microsoft.com/en-us/nuget/create-packages/creating-a-package). Also, you'll need to create what's called a Classic Personal Access Token (PAT), which you can find described elsewhere here on GH. ([Manage your PAT](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) and [authorizing a PAT for use with SSO](https://docs.github.com/en/enterprise-cloud@latest/authentication/authenticating-with-saml-single-sign-on/authorizing-a-personal-access-token-for-use-with-saml-single-sign-on).)

With all that in place, here's how I managed the details:

1. I created a folder/subdirectory on my machine to create packages I'll publish to our organization in GH. I called my folder "MakePackages".
2. Within "MakePackages" I'd create a separate folder for each DLL I wanted to publish to our org's Packages. Thus, for one of them I created a folder named DOH.DDSD.CentralRegistry, within MakePackages.
3. Within DOH.DDSD.CentralRegistry create another folder named "lib". (I stumbled on this when I tried running the one of the `nuget` commands.)
4. Within the lib folder I created another folder for the version of the .NET Framework of the DOH.DDSD.CentralRegistry.DLL. The DLL was written using .NET Framework 4.5.2, so I created a folder named `net452` within lib.
5. Then I copied DOH.DDSD.CentralRegistry.DLL into the net452 folder.
6. (I am not sure, but it might be the case that you can put different versions built on different versions of the .NET Framework within their own folders within the lib folder. I only have one version, so that's all I created.)
7. Then from a CLI prompt (in my case, PowerShell), in the net452 folder I ran the command `nuget spec DOH.DDSD.CentralRegistry.dll`. This will create a `nuspec` file in the net452 folder.
8. Move the nuspec folder up to the DOH.DDSD.CentralRegistry folder. Also change your directory in the CLI to that root level folder, in my case the DOH.DDSD.CentralRegistry folder.
9. The nuspec folder is just an XML file. I edited it with VS Code to correct some items within it. For example, one of the items is `authors`. The `nuget spec` command put my name in as the author of the DLL. This is wrong, so I replaced my name with the real author's name of the DLL. There's other items in the nuspec file, like copyright, etc. Edit the other fields as appropriate for your situation.
10. Next, you'll need to run the `nuget pack` command, in my case that was `nuget pack DOH.DDSD.CentralRegistry.dll.nuspec` command at the CLI. This creates a `nupkg` file. In my case it created a file named Doh.Ddsd.CentralRegistry.dll.1.0.0.nupkg.

### Publish your NuGet package to your organization's Packages
Now you're all set to publish your nupkg file to your organization's Packages area. This is the command I used (obfuscated appropriately):

```
nuget push  .\Doh.Ddsd.CentralRegistry.dll.1.0.0.nupkg -ApiKey ghp_<your PAT key> -src https://nuget.pkg.github.com/<Your Organization's Name>/index.json
```

### Closing comments
In writing this I know I might be "mansplaining". This is **NOT** my intent! I wrote this all up for myself in a Word document, so that I can refer to it later. Currently I'm working on publishing older DLLs to our GH organization's Packages, but once I'm done with that it might be months before I do this again.  I must document it, for myself, so that when I come back to this, I can pick up quickly what works and do it again. I hope that this is OK with you.
